[package] Correct usage of miniz API in PyTorchStreamReader

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55725 [package] Correct usage of miniz API in PyTorchStreamReader**

We were previously checking m_last_error on the miniz struct directly,
which fails to preserve internal invariants and can the leave the reader
broken in specific situations (reading a non-existent file).

Using the provided error checking API fixes this.

Differential Revision: [D27693105](https://our.internmc.facebook.com/intern/diff/D27693105)

suo committed:
[package] Throw if PyTorchStreamReader::getRecord doesn't find file

Today, there will be an internal error and the reader will be left in an
unusable state. Throw an exception but otherwise preserve internal
invariants instead.

[ghstack-poisoned]

bradleyhd commented on caffe2/serialize/inline_container_test.cc:
```diff
@@ -68,6 +68,46 @@ TEST(PyTorchStreamWriterAndReader, SaveAndLoad) {
 }
 
+TEST(PytorchStreamWriterAndReader, GetNonexistentRecordThrows) {
+  int64_t kFieldAlignment = 64L;
```
nit: unused?

bradleyhd commented on caffe2/serialize/inline_container.cc:
```diff
@@ -239,6 +246,8 @@ size_t PyTorchStreamReader::getRecordID(const std::string& name) {
 // return dataptr, size
 std::tuple<at::DataPtr, size_t> PyTorchStreamReader::getRecord(const std::string& name) {
   std::lock_guard<std::mutex> guard(reader_lock_);
+  TORCH_CHECK_VALUE(
```
is there any conern of a meaningful perf impact from always checking here? I haven't spent any time with `mz_zip_reader_locate_file` and not sure if `getRecord` is often used with high call volume or with large archives, but looks like this would call it twice per record

suo replied:
yeah reading the code more closely, we already have file presence checking in `getRecordID`, it's just the error handling doesn't work. I'll just fix it that way.

bradleyhd reviewed (approved):
LGTM, thanks for taking care of this!

suo committed:
Update on "[package] Throw if PyTorchStreamReader::getRecord doesn't find file"


Today, there will be an internal error and the reader will be left in an
unusable state. Throw an exception but otherwise preserve internal
invariants instead.

Differential Revision: [D27693105](https://our.internmc.facebook.com/intern/diff/D27693105)

[ghstack-poisoned]

suo committed:
Update on "[package] Throw if PyTorchStreamReader::getRecord doesn't find file"


Today, there will be an internal error and the reader will be left in an
unusable state. Throw an exception but otherwise preserve internal
invariants instead.

Differential Revision: [D27693105](https://our.internmc.facebook.com/intern/diff/D27693105)

[ghstack-poisoned]

suo committed:
Update on "[package] Correct usage of miniz API in PyTorchStreamReader"


We were previously checking m_last_error on the miniz struct directly,
which fails to preserve internal invariants and can the leave the reader
broken in specific situations (reading a non-existent file).

Using the provided error checking API fixes this.

Differential Revision: [D27693105](https://our.internmc.facebook.com/intern/diff/D27693105)

[ghstack-poisoned]

suo committed:
Update on "[package] Correct usage of miniz API in PyTorchStreamReader"


We were previously checking m_last_error on the miniz struct directly,
which fails to preserve internal invariants and can the leave the reader
broken in specific situations (reading a non-existent file).

Using the provided error checking API fixes this.

Differential Revision: [D27693105](https://our.internmc.facebook.com/intern/diff/D27693105)

[ghstack-poisoned]
